[JSC][32-bit] Enable fused branch compare on BBQ for 32-bit platforms

mikhailramalho opened the pull request:
#### cbb285948432c0c0fe8ef0267a94d56c6429153c
<pre>
[JSC][32-bit] Enable fused branch compare on BBQ for 32-bit platforms
<a href="https://bugs.webkit.org/show_bug.cgi?id=302472">https://bugs.webkit.org/show_bug.cgi?id=302472</a>

Reviewed by Justin Michaud.

This PR enables fused branch compare for 32-bit platforms with the goal of
reducing code size.

The 32-bit and 64-bit code paths are also unified with only two #if USE(JSVALUE64),
emitBranchI64 is the most affected but it&apos;s minimal, we only need to pass more
registers in the branch64 call.

There was also only one change in the code copied from 64-bit, which was to
handle GPR2 in the two addFusedIfCompare methods.

This feature saves -16KiB in code size of JetStream3&apos;s tfjs-wasm.js, a -3.53%
improvement:

Base total code size: 439162 bytes (429KiB)
New total code size: 423680 bytes (414KiB)

Difference (new - base): -15482 bytes (-16KiB)
Percentage change: -3.53%

* Source/JavaScriptCore/assembler/ARMv7Assembler.h:
(JSC::ARMv7Assembler::invert):
* Source/JavaScriptCore/assembler/MacroAssemblerARMv7.h:
(JSC::MacroAssemblerARMv7::armV7ConditionForHigh32):
(JSC::MacroAssemblerARMv7::armV7ConditionForLow32):
(JSC::MacroAssemblerARMv7::compare64):
(JSC::MacroAssemblerARMv7::compare32AndSetFlags):
(JSC::MacroAssemblerARMv7::branch64Impl):
(JSC::MacroAssemblerARMv7::branch64):
(JSC::MacroAssemblerARMv7::branchTest64):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::tryFoldFusedBranchCompare):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitFusedBranchCompareBranch):
(JSC::Wasm::BBQJITImpl::BBQJIT::addFusedBranchCompare):
(JSC::Wasm::BBQJITImpl::BBQJIT::addFusedIfCompare):
(JSC::Wasm::BBQJITImpl::emitBranchI32):
(JSC::Wasm::BBQJITImpl::emitBranchI64):
(JSC::Wasm::BBQJITImpl::emitBranchF32):
(JSC::Wasm::BBQJITImpl::emitBranchF64):
* Source/JavaScriptCore/wasm/WasmBBQJIT.h:
* Source/JavaScriptCore/wasm/WasmBBQJIT32_64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::tryFoldFusedBranchCompare): Deleted.
(JSC::Wasm::BBQJITImpl::BBQJIT::emitFusedBranchCompareBranch): Deleted.
(JSC::Wasm::BBQJITImpl::BBQJIT::addFusedBranchCompare): Deleted.
(JSC::Wasm::BBQJITImpl::BBQJIT::addFusedIfCompare): Deleted.
* Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::tryFoldFusedBranchCompare): Deleted.
(JSC::Wasm::BBQJITImpl::BBQJIT::emitFusedBranchCompareBranch): Deleted.
(JSC::Wasm::BBQJITImpl::BBQJIT::addFusedBranchCompare): Deleted.
(JSC::Wasm::BBQJITImpl::BBQJIT::addFusedIfCompare): Deleted.
(JSC::Wasm::BBQJITImpl::emitBranchI32): Deleted.
(JSC::Wasm::BBQJITImpl::emitBranchI64): Deleted.
(JSC::Wasm::BBQJITImpl::emitBranchF32): Deleted.
(JSC::Wasm::BBQJITImpl::emitBranchF64): Deleted.

Canonical link: <a href="https://commits.webkit.org/303254@main">https://commits.webkit.org/303254@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/41f7963c8a3dacd64e5b5ab25c17c646e6b2127c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130844 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/3167 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41803 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138272 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/83159 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/3149 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/3010 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99695 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/67870 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133790 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/2261 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/117155 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80390 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2183 "Passed tests") | [⏳ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/macOS-Tahoe-Debug-API-Tests-EWS "Waiting to run tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81526 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/122860 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110775 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/35284 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140749 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/129297 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2911 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35787 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/108212 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2957 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2623 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/108135 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27685 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2225 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113487 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/55943 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2979 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31913 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/162314 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2800 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/66371 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40490 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/3000 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2908 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/wasm/WasmBBQJIT.h:
```diff
@@ -1079,8 +1079,8 @@ class BBQJIT {
-    // FIXME: Support fused branch compare on 32-bit platforms.
-    static constexpr bool shouldFuseBranchCompare = is64Bit();
+    // Enable fused branch compare for all platforms
+    static constexpr bool shouldFuseBranchCompare = true;
```
Let's remove this in a follow-up

mikhailramalho committed:
[JSC][32-bit] Enable fused branch compare on BBQ for 32-bit platforms
https://bugs.webkit.org/show_bug.cgi?id=302472

Reviewed by Justin Michaud.

This PR enables fused branch compare for 32-bit platforms with the goal of
reducing code size.

The 32-bit and 64-bit code paths are also unified with only two #if USE(JSVALUE64),
emitBranchI64 is the most affected but it's minimal, we only need to pass more
registers in the branch64 call.

There was also only one change in the code copied from 64-bit, which was to
handle GPR2 in the two addFusedIfCompare methods.

This feature saves -16KiB in code size of JetStream3's tfjs-wasm.js, a -3.53%
improvement:

Base total code size: 439162 bytes (429KiB)
New total code size: 423680 bytes (414KiB)

Difference (new - base): -15482 bytes (-16KiB)
Percentage change: -3.53%

* Source/JavaScriptCore/assembler/ARMv7Assembler.h:
(JSC::ARMv7Assembler::invert):
* Source/JavaScriptCore/assembler/MacroAssemblerARMv7.h:
(JSC::MacroAssemblerARMv7::armV7ConditionForHigh32):
(JSC::MacroAssemblerARMv7::armV7ConditionForLow32):
(JSC::MacroAssemblerARMv7::compare64):
(JSC::MacroAssemblerARMv7::compare32AndSetFlags):
(JSC::MacroAssemblerARMv7::branch64Impl):
(JSC::MacroAssemblerARMv7::branch64):
(JSC::MacroAssemblerARMv7::branchTest64):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::tryFoldFusedBranchCompare):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitFusedBranchCompareBranch):
(JSC::Wasm::BBQJITImpl::BBQJIT::addFusedBranchCompare):
(JSC::Wasm::BBQJITImpl::BBQJIT::addFusedIfCompare):
(JSC::Wasm::BBQJITImpl::emitBranchI32):
(JSC::Wasm::BBQJITImpl::emitBranchI64):
(JSC::Wasm::BBQJITImpl::emitBranchF32):
(JSC::Wasm::BBQJITImpl::emitBranchF64):
* Source/JavaScriptCore/wasm/WasmBBQJIT.h:
* Source/JavaScriptCore/wasm/WasmBBQJIT32_64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::tryFoldFusedBranchCompare): Deleted.
(JSC::Wasm::BBQJITImpl::BBQJIT::emitFusedBranchCompareBranch): Deleted.
(JSC::Wasm::BBQJITImpl::BBQJIT::addFusedBranchCompare): Deleted.
(JSC::Wasm::BBQJITImpl::BBQJIT::addFusedIfCompare): Deleted.
* Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::tryFoldFusedBranchCompare): Deleted.
(JSC::Wasm::BBQJITImpl::BBQJIT::emitFusedBranchCompareBranch): Deleted.
(JSC::Wasm::BBQJITImpl::BBQJIT::addFusedBranchCompare): Deleted.
(JSC::Wasm::BBQJITImpl::BBQJIT::addFusedIfCompare): Deleted.
(JSC::Wasm::BBQJITImpl::emitBranchI32): Deleted.
(JSC::Wasm::BBQJITImpl::emitBranchI64): Deleted.
(JSC::Wasm::BBQJITImpl::emitBranchF32): Deleted.
(JSC::Wasm::BBQJITImpl::emitBranchF64): Deleted.

Canonical link: https://commits.webkit.org/303254@main
